### PR TITLE
[SDL 0192] Add subscribe button request

### PR DESCRIPTION
**Pull request by ychernysheva**
Fixes [FORDTCN-12166](https://adc.luxoft.com/jira/browse/FORDTCN-12166)

This PR is **ready** for review.

### Risk
This PR makes **major** API changes.

### Testing Plan
ATF test scripts, affected UTs are updated

### Summary
Add HMI subscribe button request and response

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/application.h [hunk 1]
```diff
@@ -1160,6 +1163,29 @@ class Application : public virtual InitialApplicationData,
...
+   * @brief Get map of pending button subscription requests correlation ids
+   * to button names
+   * @return pending button subscriptions map
+   */
+  virtual const ButtonSubscriptionsMap& PendingButtonSubscriptions() const = 0;
+
+  /**
+   * @brief Add  pending button subscription request
+   * @param correlation_id - correlation id of subscription request
+   * @param button_name - enum value indication button name
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/application.h:1176]:
```suggestion
   * @param button_name - enum value indicating button name
```

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 2]
```diff
@@ -320,6 +320,20 @@ class MessageHelper {
...
+   * @param button - button to be subscribed
+   * @param app_mngr - application manager
+   * @return Smart object with fulfilled request
+   */
+  static smart_objects::SmartObjectSPtr
+  CreateButtonSubscriptionHandlingRequestToHmi(
+      const uint32_t app_id,
+      const hmi_apis::Common_ButtonName::eType button,
+      const hmi_apis::FunctionID::eType function,
+      ApplicationManager& app_mngr);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/message_helper.h:335]:
```suggestion
  static smart_objects::SmartObjectSPtr
  CreateButtonSubscriptionHandlingRequestToHmi(
      const uint32_t app_id,
      const hmi_apis::Common_ButtonName::eType button_name,
      const hmi_apis::FunctionID::eType function_id,
      ApplicationManager& app_mngr);
```

File: src/components/application_manager/include/application_manager/message_helper.h [hunk 3]
```diff
@@ -320,6 +320,20 @@ class MessageHelper {
       ApplicationManager& app_mngr,
       const ButtonSubscriptions& button_subscriptions);
 
+  /**
+   * @brief Creates button subscription request to hmi
+   * @param app_id - id of application for which request should be created
+   * @param button - button to be subscribed
+   * @param app_mngr - application manager
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/message_helper.h:327]: missing description of function parameter

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  SubscribeButtonRequest(const app_mngr::commands::MessageSharedPtr& message,
+                         app_mngr::ApplicationManager& application_manager,
+                         app_mngr::rpc_service::RPCService& rpc_service,
+                         app_mngr::HMICapabilities& hmi_capabilities,
+                         policy::PolicyHandlerInterface& policy_handle);
+
+  /**
+   * @brief SubscribeButtonRequest class destructor
+   **/
+  virtual ~SubscribeButtonRequest();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h:64]: virtual is redundant here

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  /**
+   * @brief Execute command
+   **/
+  virtual void Run();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h:69]:
```suggestion
  void Run() OVERRIDE;
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  void onTimeOut() OVERRIDE;
+
+  void on_event(const application_manager::event_engine::Event& event) OVERRIDE;
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(SubscribeButtonRequest);
+
+  /**
+   * @brief Determines whether internal unsubscription must be performed
+   * and HMI UnsubscribeButton request must not be sent
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h:80]:
```suggestion
   * and HMI UnsubscribeButton request should be sent
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  SubscribeButtonResponse(const app_mngr::commands::MessageSharedPtr& message,
+                          app_mngr::ApplicationManager& application_manager,
+                          app_mngr::rpc_service::RPCService& rpc_service,
+                          app_mngr::HMICapabilities& hmi_capabilities,
+                          policy::PolicyHandlerInterface& policy_handle);
+
+  /**
+   * @brief SubscribeButtonResponse class destructor
+   **/
+  virtual ~SubscribeButtonResponse();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h:63]: virtual is redundant here

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * @brief Execute command
+   **/
+  virtual void Run();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h:68]:
```suggestion
  void Run() OVERRIDE; 
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/subscribe_button_request.cc [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+  app->AddPendingButtonSubscription(correlation_id(), button_name_);
+
+  SendRequest();
+}
+
+void SubscribeButtonRequest::onTimeOut() {
+  SDL_LOG_AUTO_TRACE();
+  application_manager_.updateRequestTimeout(
+      connection_key(), correlation_id(), 0);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/subscribe_button_request.cc:92]: @ychernysheva do we really need to set timeout to 0 here?

Comment (ychernysheva) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/subscribe_button_request.cc]: Removed in d5b3c6b

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h (hunk 8 above)
Comment (ychernysheva) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h]: Done in d5b3c6b

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h (hunk 7 above)
Comment (ychernysheva) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_response.h]: Done in d5b3c6b

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h (hunk 6 above)
Comment (ychernysheva) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h]: Done in d5b3c6b

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h (hunk 5 above)
Comment (ychernysheva) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h]: Done in d5b3c6b

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h (hunk 4 above)
Comment (ychernysheva) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/hmi/subscribe_button_request.h]: Done in d5b3c6b

File: src/components/application_manager/include/application_manager/message_helper.h (hunk 3 above)
Comment (ychernysheva) [reply src/components/application_manager/include/application_manager/message_helper.h]: Done in d5b3c6b

File: src/components/application_manager/include/application_manager/message_helper.h (hunk 2 above)
Comment (ychernysheva) [reply src/components/application_manager/include/application_manager/message_helper.h]: Done in d5b3c6b

File: src/components/application_manager/include/application_manager/application.h (hunk 1 above)
Comment (ychernysheva) [reply src/components/application_manager/include/application_manager/application.h]: Done in d5b3c6b

**Review by AKalinich-Luxoft — APPROVED** [review]

**Review by RomanReznichenkoLuxoft — APPROVED** [review]

**Pull request merged**